webextension/background.js

```diff
@@ -0,0 +1,1006 @@
+          browser.experiments.aboutConfigPrefs.setString("reportEndpoint", this._reportLanding);
+        }
+
+        this._uiVariant = studyInfo.variation.name;
+
+        // Testers may use an about:config flag to toggle the UI variant.
+        // They may also use an invalid value to reset our config.
+        if (this._testingMode && variationPref !== undefined) {
+          if (this._onVariationPrefChanged(variationPref)) {
+            delete otherPrefs.uiVariant;
+          } else {
+            otherPrefs = {};
+          }
+        }
+
+        // The list of domains to check needs special handling, as the list may
+        // change when the addon updates, and we must keep the new ones and
+        // remove the no-longer-interesting ones, while retaining whether the
+        // user has been prompted for any of the final domains.
+        if ("domainsToCheck" in otherPrefs) {
+          let foundChange = false;
+          const oldDomains = otherPrefs.domainsToCheck;
+          for (const domain of Object.keys(this._domainsToCheck)) {
+            if (domain in oldDomains) {
+              this._domainsToCheck[domain] = oldDomains[domain];
+              foundChange = true;
+            }
+          }
+          if (foundChange) {
+            this.save({domainsToCheck: this._domainsToCheck});
+          }
+          delete otherPrefs.domainsToCheck;
+        }
+
+        // The rest of the values can just be set on the object as-is, since
+        // we will have written them out with a valid value to begin with.
+        for (const [name, value] of Object.entries(otherPrefs)) {
+          this[`_${name}`] = value;
+        }
+
+        // If a valid UI variant has not otherwise been chosen yet, select one now.
+        if (!this._uiVariant) {
+          this._selectRandomUIVariant();
+        }
+
+        this._loaded = true;
+      });
+    }
+
+    save(options) {
+      const promises = [];
+      if (Object.keys(options).length) {
+        promises.push(browser.storage.local.set(options));
+      }
+      return Promise.all(promises);
+    }
+
+    onUserPrompted(url) {
+      const domain = this.findDomainMatch(new URL(url).host);
+      if (!domain) {
+        return;
+      }
+      const now = Date.now();
+      this._lastPromptTime = now;
+      this._domainsToCheck[domain].lastPromptTime = now;
+      this.save({
+        lastPromptTime: now,
+        domainsToCheck: this._domainsToCheck,
+      });
+    }
+
+    findDomainMatch(domain) {
+      for (const candidate of Object.keys(this._domainsToCheck)) {
+        if (domain.endsWith(candidate)) {
+          return candidate;
+        }
+      }
+      return undefined;
+    }
+
+    isPromptableURL(url) {
+      if (!(url instanceof URL)) {
+        try {
+          url = new URL(url);
+        } catch (_) {
+          return false;
+        }
+      }
+      return ["http:", "https:"].includes(url.protocol);
+    }
+
+    shouldPromptUserNow(domain) {
```

Comment:
I'm confused by this function, is the comment out of date with the implementation?

Reply:
Yes, this is outdated and iffy. I'll reimplement it to be clearer and to check for the at-most-5-times case.